File: python/src/renderer.py
"""
Terminal renderer for the racing game
Demonstrates: F-strings, ANSI escape codes, context managers, type hints
"""

import os
import sys
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from game import Game

from config import TRACK_LENGTH, NUM_LANES, LANE_WIDTH


class Renderer:
    """
    Renderer class for terminal-based visualization
    Demonstrates: Type hints, f-strings, ANSI codes, context manager
    """

    def __init__(self) -> None:
        """Initialize the renderer"""
        self.track_length: int = TRACK_LENGTH
        self.num_lanes: int = NUM_LANES
        self.lane_width: int = LANE_WIDTH

    def clear(self) -> None:
        """
        Clear the terminal screen
        Demonstrates: OS-specific commands, ANSI escape codes
        """
        # ANSI escape sequence to clear screen and move cursor to top
        sys.stdout.write('\033[2J\033[H')
        sys.stdout.flush()

    def hide_cursor(self) -> None:
        """
        Hide cursor for cleaner rendering
        Demonstrates: ANSI escape codes
        """
        sys.stdout.write('\033[?25l')
        sys.stdout.flush()

    def show_cursor(self) -> None:
        """
        Show cursor
        Demonstrates: ANSI escape codes
        """
        sys.stdout.write('\033[?25h')
        sys.stdout.flush()

    def draw_border(self) -> str:
        """
        Draw a horizontal border line
        Demonstrates: String multiplication, f-strings

        Returns:
            Border string
        """
        width = self.lane_width * self.num_lanes + 2
        return '┌' + '─' * width + '┐'

    def draw_header(self) -> str:
        """
        Draw the header
        Demonstrates: Multi-line f-strings, list comprehension, str.join

        Returns:
            Header string
        """
        title = 'VERTICAL LANE RACER v1.0'
        lane_labels = ''.join(f' {i}  ' for i in range(self.num_lanes))

        width = self.lane_width * self.num_lanes + 2

        return '\n'.join([
            self.draw_border(),
            f'│  {title:<{self.lane_width * self.num_lanes - 2}}  │',
            f'│  Lane: {lane_labels}│',
            '├' + '─' * (self.lane_width * self.num_lanes + 2) + '┤',
        ])

    def build_track_grid(self, game: 'Game') -> list[str]:
        """
        Build track grid with cars and obstacles
        Demonstrates: List comprehension, nested loops, type hints

        Args:
            game: Game instance

        Returns:
            List of track lines
        """
        # Create empty grid (each cell is one character position)
        grid: list[list[str | None]] = [
            [None for _ in range(self.num_lanes)]
            for _ in range(self.track_length)
        ]

        # Place obstacles on grid
        for obstacle in game.obstacles:
            row = int(obstacle.position)
            if 0 <= row < self.track_length:
                grid[row][obstacle.lane] = obstacle.symbol

        # Place cars on grid (cars overlay obstacles)
        for car in game.cars:
            if car.status != 'finished':
                row = int(car.position)
                if 0 <= row < self.track_length:
                    grid[row][car.lane] = car.symbol

        # Convert grid to string lines
        lines: list[str] = []
        for row in range(self.track_length):
            line = '│  '
            for lane in range(self.num_lanes):
                cell = grid[row][lane]
                if cell:
                    line += f'{cell}   '
                else:
                    line += '|   '
            line += ' │'
            lines.append(line)

        return lines

    def draw_footer(self) -> str:
        """
        Draw the footer with finish line
        Demonstrates: String multiplication

        Returns:
            Footer string
        """
        width = self.lane_width * self.num_lanes + 2
        return '├' + '─' * width + '┤'

    def draw_stats(self, game: 'Game') -> str:
        """
        Draw car statistics
        Demonstrates: List comprehension, dict literal, match/case, f-strings

        Args:
            game: Game instance

        Returns:
            Stats string
        """
        stats = game.get_stats()

        # Map status to emoji using match/case (Python 3.10+)
        def get_status_emoji(status: str) -> str:
            match status:
                case 'active':
                    return '🏃'
                case 'crashed':
                    return '💥'
                case 'finished':
                    return '🏁'
                case _:
                    return '?'

        # Build car info lines using list comprehension
        car_lines = [
            f'│ {car.symbol} Car {car.id}: Lane {car.lane}, '
            f'Pos {car.position:>4.1f}, '
            f'Speed {car.speed:.1f} {get_status_emoji(car.status):<2} │'
            for car in game.cars
        ]

        tick_line = (
            f"│ Tick: {stats['tick']:>3} "
            f"| Active: {stats['active_cars']} | Crashed: {stats['crashed_cars']} "
            f"| Finished: {stats['finished_cars']}"
        ).ljust(self.lane_width * self.num_lanes + 2) + ' │'

        border = '├' + '─' * (self.lane_width * self.num_lanes + 2) + '┤'

        return '\n'.join([
            *car_lines,
            border,
            tick_line,
        ])

    def draw_bottom_border(self) -> str:
        """
        Draw the bottom border
        Demonstrates: String multiplication

        Returns:
            Bottom border string
        """
        width = self.lane_width * self.num_lanes + 2
        return '└' + '─' * width + '┘'

    def render(self, game: 'Game') -> None:
        """
        Render the complete game frame
        Demonstrates: Method composition, list unpacking, str.join

        Args:
            game: Game instance
        """
        self.clear()

        output = '\n'.join([
            self.draw_header(),
            *self.build_track_grid(game),
            self.draw_footer(),
            self.draw_stats(game),
            self.draw_bottom_border(),
        ])

        print(output)

    def show_results(self, game: 'Game') -> None:
        """
        Show final results screen
        Demonstrates: Multi-line f-strings, optional types

        Args:
            game: Game instance
        """
        self.clear()

        winner = game.get_winner()
        stats = game.get_stats()

        print('\n' + '═' * 50)
        print('🏁  RACE FINISHED  🏁'.center(50))
        print('═' * 50 + '\n')

        if winner:
            print(f'🎉 WINNER: Car {winner.id} {winner.symbol}')
            print(f'   Final Position: {winner.position:.2f}')
            print(f'   Final Lane: {winner.lane}')
            print(f'   Speed: {winner.speed:.2f}')
        else:
            print('❌ NO WINNER - All cars crashed!')

        print('\n' + '─' * 50)
        print('📊 FINAL STATISTICS\n')
        print(f"   Total Ticks: {stats['tick']}")
        print(f"   Active Cars: {stats['active_cars']}")
        print(f"   Crashed Cars: {stats['crashed_cars']}")
        print(f"   Finished Cars: {stats['finished_cars']}")

        print('\n🚗 CAR DETAILS\n')
        for car in game.cars:
            status = car.status.upper()
            print(f'   {car.symbol} Car {car.id}: {status} at position {car.position:.2f}')

        print('\n' + '═' * 50 + '\n')

        self.show_cursor()

    def cleanup(self) -> None:
        """
        Clean up renderer (show cursor, etc.)
        Demonstrates: Cleanup pattern
        """
        self.show_cursor()

    def __enter__(self):
        """
        Context manager entry
        Demonstrates: Context manager protocol (__enter__/__exit__)
        """
        self.hide_cursor()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """
        Context manager exit
        Ensures cursor is shown even if an error occurs
        """
        self.cleanup()
        return False  # Don't suppress exceptions

File: javascript/src/renderer.js
/**
 * Terminal renderer for the racing game
 * Demonstrates: Template literals, ANSI escape codes, string manipulation
 */

import * as readline from 'readline';
import { TRACK_LENGTH, NUM_LANES, LANE_WIDTH } from './config.js';

/**
 * Renderer class for terminal-based visualization
 * Demonstrates: ES6 classes, template literals, ANSI codes
 */
export class Renderer {
  constructor() {
    this.trackLength = TRACK_LENGTH;
    this.numLanes = NUM_LANES;
    this.laneWidth = LANE_WIDTH;
  }

  /**
   * Clear the terminal screen
   * Demonstrates: ANSI escape codes, process.stdout manipulation
   */
  clear() {
    // ANSI escape sequence to clear screen and move cursor to top
    process.stdout.write('\x1Bc');
    // Alternative: console.clear();
  }

  /**
   * Move cursor to specific position
   * Demonstrates: readline module usage
   * @param {number} x - Column position
   * @param {number} y - Row position
   */
  moveCursor(x, y) {
    readline.cursorTo(process.stdout, x, y);
  }

  /**
   * Hide cursor for cleaner rendering
   * Demonstrates: ANSI escape codes
   */
  hideCursor() {
    process.stdout.write('\x1B[?25l');
  }

  /**
   * Show cursor
   * Demonstrates: ANSI escape codes
   */
  showCursor() {
    process.stdout.write('\x1B[?25h');
  }

  /**
   * Draw a horizontal border line
   * Demonstrates: String.repeat, template literals
   * @returns {string}
   */
  drawBorder() {
    const width = this.laneWidth * this.numLanes + 2;
    return '┌' + '─'.repeat(width) + '┐';
  }

  /**
   * Draw the header
   * Demonstrates: Template literals, string padding
   * @returns {string}
   */
  drawHeader() {
    const title = 'VERTICAL LANE RACER v1.0';
    const laneLabels = Array.from({ length: this.numLanes }, (_, i) => i)
      .map(i => ` ${i}  `)
      .join('');

    return [
      this.drawBorder(),
      `│  ${title.padEnd(this.laneWidth * this.numLanes - 2)}  │`,
      `│  Lane: ${laneLabels}│`,
      '├' + '─'.repeat(this.laneWidth * this.numLanes + 2) + '┤',
    ].join('\n');
  }

  /**
   * Build track grid with cars and obstacles
   * Demonstrates: 2D array creation, Map usage, complex logic
   * @param {Game} game - Game instance
   * @returns {string[]} - Array of track lines
   */
  buildTrackGrid(game) {
    // Create empty grid (each cell is one character position)
    const grid = Array.from({ length: this.trackLength }, () =>
      Array(this.numLanes).fill(null)
    );

    // Place obstacles on grid
    for (const obstacle of game.obstacles) {
      const row = Math.floor(obstacle.position);
      if (row >= 0 && row < this.trackLength) {
        grid[row][obstacle.lane] = obstacle.symbol;
      }
    }

    // Place cars on grid (cars overlay obstacles)
    for (const car of game.cars) {
      if (car.status !== 'finished') {
        const row = Math.floor(car.position);
        if (row >= 0 && row < this.trackLength) {
          grid[row][car.lane] = car.symbol;
        }
      }
    }

    // Convert grid to string lines
    const lines = [];
    for (let row = 0; row < this.trackLength; row++) {
      let line = '│  ';
      for (let lane = 0; lane < this.numLanes; lane++) {
        const cell = grid[row][lane];
        if (cell) {
          line += `${cell}   `;
        } else {
          line += '|   ';
        }
      }
      line += ' │';
      lines.push(line);
    }

    return lines;
  }

  /**
   * Draw the footer with finish line
   * Demonstrates: Template literals, string manipulation
   * @returns {string}
   */
  drawFooter() {
    const width = this.laneWidth * this.numLanes + 2;
    return '├' + '─'.repeat(width) + '┤';
  }

  /**
   * Draw car statistics
   * Demonstrates: Array.map, template literals, destructuring
   * @param {Game} game - Game instance
   * @returns {string}
   */
  drawStats(game) {
    const stats = game.getStats();
    const carLines = game.cars.map(car => {
      const statusEmoji = {
        active: '🏃',
        crashed: '💥',
        finished: '🏁',
      }[car.status] || '?';

      return `│ ${car.symbol} Car ${car.id}: Lane ${car.lane}, ` +
        `Pos ${car.position.toFixed(1).padStart(4)}, ` +
        `Speed ${car.speed.toFixed(1)} ${statusEmoji.padEnd(2)} │`;
    });

    const tickLine = `│ Tick: ${stats.tick.toString().padStart(3)} ` +
      `| Active: ${stats.activeCars} | Crashed: ${stats.crashedCars} ` +
      `| Finished: ${stats.finishedCars}`.padEnd(this.laneWidth * this.numLanes - 12) + ' │';

    return [
      ...carLines,
      '├' + '─'.repeat(this.laneWidth * this.numLanes + 2) + '┤',
      tickLine,
    ].join('\n');
  }

  /**
   * Draw the bottom border
   * Demonstrates: Template literals
   * @returns {string}
   */
  drawBottomBorder() {
    const width = this.laneWidth * this.numLanes + 2;
    return '└' + '─'.repeat(width) + '┘';
  }

  /**
   * Render the complete game frame
   * Demonstrates: Method composition, template literals, string concatenation
   * @param {Game} game - Game instance
   */
  render(game) {
    this.clear();

    const output = [
      this.drawHeader(),
      ...this.buildTrackGrid(game),
      this.drawFooter(),
      this.drawStats(game),
      this.drawBottomBorder(),
    ].join('\n');

    console.log(output);
  }

  /**
   * Show final results screen
   * Demonstrates: Template literals, optional chaining
   * @param {Game} game - Game instance
   */
  showResults(game) {
    this.clear();

    const winner = game.getWinner();
    const stats = game.getStats();

    console.log('\n' + '═'.repeat(50));
    console.log('🏁  RACE FINISHED  🏁'.padStart(32));
    console.log('═'.repeat(50) + '\n');

    if (winner) {
      console.log(`🎉 WINNER: Car ${winner.id} ${winner.symbol}`);
      console.log(`   Final Position: ${winner.position.toFixed(2)}`);
      console.log(`   Final Lane: ${winner.lane}`);
      console.log(`   Speed: ${winner.speed.toFixed(2)}`);
    } else {
      console.log('❌ NO WINNER - All cars crashed!');
    }

    console.log('\n' + '─'.repeat(50));
    console.log('📊 FINAL STATISTICS\n');
    console.log(`   Total Ticks: ${stats.tick}`);
    console.log(`   Active Cars: ${stats.activeCars}`);
    console.log(`   Crashed Cars: ${stats.crashedCars}`);
    console.log(`   Finished Cars: ${stats.finishedCars}`);

    console.log('\n🚗 CAR DETAILS\n');
    for (const car of game.cars) {
      const status = car.status.toUpperCase();
      console.log(`   ${car.symbol} Car ${car.id}: ${status} at position ${car.position.toFixed(2)}`);
    }

    console.log('\n' + '═'.repeat(50) + '\n');

    this.showCursor();
  }

  /**
   * Clean up renderer (show cursor, etc.)
   * Demonstrates: Cleanup pattern
   */
  cleanup() {
    this.showCursor();
  }
}
